frontend/src/metabase-lib/lib/queries/NativeQuery.js

```diff
@@ -44,6 +44,19 @@ export const NATIVE_QUERY_TEMPLATE: NativeDatasetQuery = {
   },
 };
 
+function cardTagCardId(name) {
+  const reCardTag = /^#([0-9]*)$/g;
+  const match = reCardTag.exec(name);
+  if (match !== null && match[1].length > 0) {
```

Comment:
I'd probably do

```
const match = name.match(/^#([0-9]+)$/);
if (match) {
  return parseInt(match[1])
}
```

I avoid `exec` for the reason I think you discovered (it's not stateless)

Reply:
Also we allow whitespace in template tags, so maybe change the regex to `/^#\s*([0-9]+)\s*$/`, move it to a `const CARD_TAG_REGEX` outside the function and use in both places (ideally in `replaceCardId` too but that's a little trickier)